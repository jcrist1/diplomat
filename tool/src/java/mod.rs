use std::collections::HashSet;
use std::fs::File;
use std::io::Write;
use std::path::Path;

use diplomat_core::hir::TypeContext;

use crate::c2;
use crate::common::FileMap;

const TMP_C_DIR: &str = "tmp";
const LIBRARY: &str = "somelib"; // todo: build from conf. Ensure that name is not the same as any
                                 // type
const GROUP: &str = "dev.diplomattest"; // todo: config
const TMP_LIB_NAME: &str = "dev/diplomattest/somelib"; // todo: build from conf
const JAVA_DIR: &str = "src/main/java/";

pub fn run(
    tcx: &TypeContext,
    _conf_path: Option<&Path>,
    out_folder: &Path,
) -> std::io::Result<FileMap> {
    let files = FileMap::default();
    let mut context = c2::CContext::new(tcx, files, false);
    context.run();

    let errors = context.errors.take_all();

    if !errors.is_empty() {
        eprintln!("Found errors when generating c  code");
        for error in errors {
            eprintln!("\t{}: {}", error.0, error.1);
        }
    }

    let out_files = context.files.take_files();

    let tmp_path = out_folder.join(TMP_C_DIR);
    std::fs::create_dir(&tmp_path)?;
    let mut include_files = HashSet::new();
    for (subpath, text) in out_files {
        let out_path = tmp_path.join(&subpath);
        if !subpath.ends_with(".d.h") && subpath.ends_with(".h") {
            include_files.insert(subpath);
        }
        let parent = out_path
            .parent()
            .expect("Cannot create files at top level dir /");
        std::fs::create_dir_all(parent)?;
        let mut out_file = File::create(&out_path)?;
        out_file.write_all(text.as_bytes())?;
    }

    let lib_path = tmp_path.join(format!("{LIBRARY}.h"));

    let mut lib_file = File::create(&lib_path)?;
    for include in include_files {
        writeln!(lib_file, "#include \"{include}\"")?;
    }

    // jextract \
    //   --include-dir /path/to/mylib/include \
    //   --output src \
    //   --target-package org.jextract.mylib \
    //   --library mylib \
    //   /path/to/mylib/include/mylib.h

    let package = format!("{GROUP}.{LIBRARY}.ntv");
    let mut command = std::process::Command::new("jextract");
    command
        .arg("--include-dir")
        .arg(&tmp_path)
        .arg("--output")
        .arg(out_folder)
        .arg("--target-package")
        .arg(package)
        .arg("--library")
        .arg(LIBRARY)
        .arg(lib_path);

    println!("Running: {:?}", command);

    // todo: delete directory

    match command.output() {
        Err(err) => match err.kind() {
            std::io::ErrorKind::NotFound => {
                eprintln!("Check that jextract is in your path and all directories exist. See https://github.com/openjdk/jextract/blob/5715737be0a1a9de24cce3ee7190881cfc8b1350/doc/GUIDE.md");
                return Err(err);
            }
            _ => return Err(err),
        },
        Ok(ok) => {
            let stdout = String::from_utf8_lossy(&ok.stdout);
            println!("Output from jextract:\n{stdout}");

            let stderr = String::from_utf8_lossy(&ok.stderr);
            println!("Std Err from jextract:\n{stderr}");
        }
    }

    let files = FileMap::default();
    Ok(files)
}
