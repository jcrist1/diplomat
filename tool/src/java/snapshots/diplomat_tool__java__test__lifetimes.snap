---
source: tool/src/java/mod.rs
expression: res
---
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;
import java.lang.foreign.SegmentAllocator;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class BorrowedFields {
    String a;
    String b;
    String c;
    

    SegmentAllocator arena;
    List<Object> selfEdges = List.of();
    List<Object> aEdges = List.of();
    

    private BorrowedFields(SegmentAllocator arena) {
        this.arena = arena;
    }

    BorrowedFields(SegmentAllocator arena, MemorySegment structSegment, List<Object> aEdges) {
        this.arena = arena;
        this.selfEdges = selfEdges;
        this.aEdges = aEdges;
        

        this.a = SliceUtils.readUtf16(dev.diplomattest.somelib.ntv.BorrowedFields.a(structSegment));
        this.b = SliceUtils.readUtf8(dev.diplomattest.somelib.ntv.BorrowedFields.b(structSegment));
        this.c = SliceUtils.readUtf8(dev.diplomattest.somelib.ntv.BorrowedFields.c(structSegment));
        

    }

    MemorySegment toNative(SegmentAllocator arena) {
        var returnVal = dev.diplomattest.somelib.ntv.BorrowedFields.allocate(arena);
        
        dev.diplomattest.somelib.ntv.BorrowedFields.a(returnVal, SliceUtils.strToUtf16Slice(arena, this.a));
        dev.diplomattest.somelib.ntv.BorrowedFields.b(returnVal, SliceUtils.strToUtf8Slice(arena, this.b));
        dev.diplomattest.somelib.ntv.BorrowedFields.c(returnVal, SliceUtils.strToUtf8Slice(arena, this.c));
        

        return returnVal;

    }
    
}


============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;
import java.lang.foreign.SegmentAllocator;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class BorrowedFieldsReturning {
    String bytes;
    

    SegmentAllocator arena;
    List<Object> selfEdges = List.of();
    List<Object> aEdges = List.of();
    

    private BorrowedFieldsReturning(SegmentAllocator arena) {
        this.arena = arena;
    }

    BorrowedFieldsReturning(SegmentAllocator arena, MemorySegment structSegment, List<Object> aEdges) {
        this.arena = arena;
        this.selfEdges = selfEdges;
        this.aEdges = aEdges;
        

        this.bytes = SliceUtils.readUtf8(dev.diplomattest.somelib.ntv.BorrowedFieldsReturning.bytes(structSegment));
        

    }

    MemorySegment toNative(SegmentAllocator arena) {
        var returnVal = dev.diplomattest.somelib.ntv.BorrowedFieldsReturning.allocate(arena);
        
        dev.diplomattest.somelib.ntv.BorrowedFieldsReturning.bytes(returnVal, SliceUtils.strToUtf8Slice(arena, this.bytes));
        

        return returnVal;

    }
    
}


============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;
import java.lang.foreign.SegmentAllocator;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class BorrowedFieldsWithBounds {
    String fieldA;
    String fieldB;
    String fieldC;
    

    SegmentAllocator arena;
    List<Object> selfEdges = List.of();
    List<Object> aEdges = List.of();
    List<Object> bEdges = List.of();
    List<Object> cEdges = List.of();
    

    private BorrowedFieldsWithBounds(SegmentAllocator arena) {
        this.arena = arena;
    }

    BorrowedFieldsWithBounds(SegmentAllocator arena, MemorySegment structSegment, List<Object> aEdges, List<Object> bEdges, List<Object> cEdges) {
        this.arena = arena;
        this.selfEdges = selfEdges;
        this.aEdges = aEdges;
        this.bEdges = bEdges;
        this.cEdges = cEdges;
        

        this.fieldA = SliceUtils.readUtf16(dev.diplomattest.somelib.ntv.BorrowedFieldsWithBounds.field_a(structSegment));
        this.fieldB = SliceUtils.readUtf8(dev.diplomattest.somelib.ntv.BorrowedFieldsWithBounds.field_b(structSegment));
        this.fieldC = SliceUtils.readUtf8(dev.diplomattest.somelib.ntv.BorrowedFieldsWithBounds.field_c(structSegment));
        

    }

    MemorySegment toNative(SegmentAllocator arena) {
        var returnVal = dev.diplomattest.somelib.ntv.BorrowedFieldsWithBounds.allocate(arena);
        
        dev.diplomattest.somelib.ntv.BorrowedFieldsWithBounds.field_a(returnVal, SliceUtils.strToUtf16Slice(arena, this.fieldA));
        dev.diplomattest.somelib.ntv.BorrowedFieldsWithBounds.field_b(returnVal, SliceUtils.strToUtf8Slice(arena, this.fieldB));
        dev.diplomattest.somelib.ntv.BorrowedFieldsWithBounds.field_c(returnVal, SliceUtils.strToUtf8Slice(arena, this.fieldC));
        

        return returnVal;

    }
    
}


============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Bar {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    List<Object> selfEdges = List.of();
    List<Object> bEdges = List.of();
    List<Object> aEdges = List.of();
    

    static class BarCleaner implements Runnable {

        MemorySegment segment;
        BarCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Bar_destroy(this.segment);
        }
    }

    Bar() {}
    Bar(MemorySegment handle, List<Object> selfEdges, List<Object> bEdges, List<Object> aEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        this.bEdges = bEdges;
        this.aEdges = aEdges;
        

    }
    
    
}
============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Foo {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    List<Object> selfEdges = List.of();
    List<Object> aEdges = List.of();
    

    static class FooCleaner implements Runnable {

        MemorySegment segment;
        FooCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Foo_destroy(this.segment);
        }
    }

    Foo() {}
    Foo(MemorySegment handle, List<Object> selfEdges, List<Object> aEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        this.aEdges = aEdges;
        

    }
    
    public static Foo new_(String x) {
        
        try (var arena = Arena.ofConfined()) {
            var xData= arena.allocateFrom(x, StandardCharsets.UTF_8);
            var xLen = xData.byteSize() - 1;
            var xView = DiplomatStringView.allocate(arena);
            DiplomatStringView.len(xView, xLen);
            DiplomatStringView.data(xView, xData);
            var nativeVal = somelib_h.Foo_new(xView);
            List<Object> selfEdges = List.of();
            
            
            
            List<Object> aEdges = List.of(x);
            var returnVal = new Foo(nativeVal, selfEdges, aEdges);
            var cleaner = new Foo.FooCleaner(nativeVal);
            returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
            return returnVal;
        }
    }
    
    public static Foo extractFromFields(BorrowedFields fields) {
        
        var fieldsNative = fields.toNative(fields.arena);
        var nativeVal = somelib_h.Foo_extract_from_fields(fieldsNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> aEdges = List.of(fields);
        var returnVal = new Foo(nativeVal, selfEdges, aEdges);
        var cleaner = new Foo.FooCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static Foo extractFromBounds(BorrowedFieldsWithBounds bounds,String anotherString) {
        
        try (var arena = Arena.ofConfined()) {
            var boundsNative = bounds.toNative(bounds.arena);
            var anotherStringData= arena.allocateFrom(anotherString, StandardCharsets.UTF_8);
            var anotherStringLen = anotherStringData.byteSize() - 1;
            var anotherStringView = DiplomatStringView.allocate(arena);
            DiplomatStringView.len(anotherStringView, anotherStringLen);
            DiplomatStringView.data(anotherStringView, anotherStringData);
            var nativeVal = somelib_h.Foo_extract_from_bounds(boundsNative, anotherStringView);
            List<Object> selfEdges = List.of();
            
            
            
            List<Object> aEdges = List.of(bounds, bounds, anotherString);
            var returnVal = new Foo(nativeVal, selfEdges, aEdges);
            var cleaner = new Foo.FooCleaner(nativeVal);
            returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
            return returnVal;
        }
    }
    
    
    public Bar getBar() {
        
        
        var nativeVal = somelib_h.Foo_get_bar(internal);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> bEdges = List.of(this);
        List<Object> aEdges = List.of(this);
        var returnVal = new Bar(nativeVal, selfEdges, bEdges, aEdges);
        var cleaner = new Bar.BarCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public BorrowedFieldsReturning asReturning() {
        
        var returnArena = (SegmentAllocator) Arena.ofAuto();
        
        var nativeVal = somelib_h.Foo_as_returning(returnArena, internal);
        
        List<Object> aEdges = List.of(this);
        
        var returnVal = new BorrowedFieldsReturning(returnArena, nativeVal, aEdges);
        return returnVal;
    }
    
}
============================
